dev/com.ibm.ws.kernel.boot.core/src/com/ibm/ws/kernel/boot/internal/BootstrapManifest.java

```diff
@@ -150,6 +152,75 @@ public String getBundleVersion() {
         return manifestAttributes.getValue(BUNDLE_VERSION);
     }
 
+    @SuppressWarnings("unchecked")
+    private String calculateSystemPackages(String javaVersion) {
+        try {
+            javaVersion = javaVersion.split("\\.")[0];
+            int version = Integer.parseInt(javaVersion);
+            if (version < 9) {
+                return null;
+            }
+            List<String> packages = new ArrayList<>();
+            Method classGetModule = Class.class.getMethod("getModule"); //$NON-NLS-1$
+            Object thisModule = classGetModule.invoke(getClass());
+            Class<?> moduleLayerClass = Class.forName("java.lang.ModuleLayer"); //$NON-NLS-1$
+            Method boot = moduleLayerClass.getMethod("boot"); //$NON-NLS-1$
+            Method modules = moduleLayerClass.getMethod("modules"); //$NON-NLS-1$
+            Class<?> moduleClass = Class.forName("java.lang.Module"); //$NON-NLS-1$
+            Method getDescriptor = moduleClass.getMethod("getDescriptor"); //$NON-NLS-1$
+            Class<?> moduleDescriptorClass = Class.forName("java.lang.module.ModuleDescriptor"); //$NON-NLS-1$
+            Method exports = moduleDescriptorClass.getMethod("exports"); //$NON-NLS-1$
+            Method isAutomatic = moduleDescriptorClass.getMethod("isAutomatic"); //$NON-NLS-1$
+            Method packagesMethod = moduleDescriptorClass.getMethod("packages"); //$NON-NLS-1$
+            Class<?> exportsClass = Class.forName("java.lang.module.ModuleDescriptor$Exports"); //$NON-NLS-1$
+            Method isQualified = exportsClass.getMethod("isQualified"); //$NON-NLS-1$
+            Method source = exportsClass.getMethod("source"); //$NON-NLS-1$
+
+            Object bootLayer = boot.invoke(null);
+            Set<?> bootModules = (Set<?>) modules.invoke(bootLayer);
```

Comment:
Since we are using reflection and can't use generic types It would be useful to add a comment to indicate what the types are. I also feel like it would be useful to have comments indicating what the non-reflective code equivalent would be, but I think that risk getting out of sync so probably isn't worth it.

Reply:
I did this anyway.